### PR TITLE
feat: openapi version 3 support non-auth changes

### PR DESCRIPTION
**Pull request by praneetap**
*Issue #, if available:*
#191 #627 
*Description of changes:*
Adding OpenApi 3.0 support for non-auth use cases. This adds a global/api flag called OpenApiVersion, which when specified - 
1. with value 2.0 or 3.0, fixes the issue #191 
2. With value 3.0, changes the swagger to openapi 3.0 format. 

*Description of how you validated changes:*
Added tests, and end to end tests.
*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by praneetap** [comment]
#191

**Comment by praneetap** [comment]
#627

File: samtranslator/model/api/api_generator.py [hunk 1]
```diff
@@ -93,6 +96,10 @@ def _construct_rest_api(self):
             raise InvalidResourceException(self.logical_id,
                                            "Specify either 'DefinitionUri' or 'DefinitionBody' property and not both")
 
+        if self.open_api_version and self.open_api_version not in OpenApiVersionsSupported:
```
Comment (keetonian) [inline samtranslator/model/api/api_generator.py]: So I think people often use the patch version for OpenApi as well (ie '3.0.0' or '3.0.1'); I think we just want to verify that they're using some version of OpenApi 3 and not care about the minor/patch versions.

File: samtranslator/model/api/api_generator.py [hunk 2]
```diff
@@ -147,7 +154,8 @@ def _construct_deployment(self, rest_api):
         deployment = ApiGatewayDeployment(self.logical_id + 'Deployment',
                                           attributes=self.passthrough_resource_attributes)
         deployment.RestApiId = rest_api.get_runtime_attr('rest_api_id')
-        deployment.StageName = 'Stage'
+        if not self.open_api_version:
+            deployment.StageName = 'Stage'
```
Comment (keetonian) [inline samtranslator/model/api/api_generator.py]: I am so glad we are finally getting rid of this.

I proposed also getting rid of it if we detect that they're using OpenApi 3.x in their swagger (but don't have the OpenApi flag set), are we going ahead with that idea or not?

File: samtranslator/model/apigateway.py [hunk 3]
```diff
@@ -20,7 +20,8 @@ class ApiGatewayRestApi(Resource):
             'Parameters': PropertyType(False, is_type(dict)),
             'EndpointConfiguration': PropertyType(False, is_type(dict)),
             "BinaryMediaTypes": PropertyType(False, is_type(list)),
-            "MinimumCompressionSize": PropertyType(False, is_type(int))
+            "MinimumCompressionSize": PropertyType(False, is_type(int)),
+            'OpenApiVersion': PropertyType(False, is_str())
```
Comment (keetonian) [inline samtranslator/model/apigateway.py]: Remove this line. This is the ApiGateway resource (maps directly to the CFN equivalent). The OpenApiVersion flag only needs to be added to the `Serverless::Api` resource found in the `sam_resources.py` file.

File: samtranslator/model/apigateway.py [hunk 4]
```diff
@@ -39,7 +40,7 @@ class ApiGatewayStage(Resource):
             'DeploymentId': PropertyType(True, is_str()),
             'Description': PropertyType(False, is_str()),
             'RestApiId': PropertyType(True, is_str()),
-            'StageName': PropertyType(True, one_of(is_str(), is_type(dict))),
+            'StageName': PropertyType(False, one_of(is_str(), is_type(dict))),
```
Comment (keetonian) [inline samtranslator/model/apigateway.py]: Why did you change this to `False`?

File: samtranslator/swagger/swagger.py [hunk 5]
```diff
@@ -554,10 +554,13 @@ def is_valid(data):
...
         """
-        return bool(data) and \
-            isinstance(data, dict) and \
-            bool(data.get("swagger")) and \
-            isinstance(data.get('paths'), dict)
+        if bool(data) and isinstance(data, dict) and isinstance(data.get('paths'), dict):
+            if bool(data.get("swagger")):
+                return True
+            elif bool(data.get("openapi")):
+                return "3.0" in data["openapi"]
```
Comment (keetonian) [inline samtranslator/swagger/swagger.py]: Same callout about OpenAPI versions (3.0.0 should be supported, I think)

Comment (praneetap) [reply samtranslator/swagger/swagger.py]: 3.0.0 is supported, I have a test for it..

File: samtranslator/model/apigateway.py (hunk 4 above)
Comment (praneetap) [reply samtranslator/model/apigateway.py]: I followed Sanath's previous change for stage stage fix. If this is required I will have to set it, which is not what we want.

File: samtranslator/model/api/api_generator.py (hunk 1 above)
Comment (praneetap) [reply samtranslator/model/api/api_generator.py]: Do you recommend adding a regex here?

File: samtranslator/model/api/api_generator.py (hunk 2 above)
Comment (praneetap) [reply samtranslator/model/api/api_generator.py]: Unfortunately there are cases where we don't read through the swagger document (definitionUri) so I don't think that is a consistent experience.

Comment (brettstack) [reply samtranslator/model/api/api_generator.py]: Could you elaborate? We should be able to get rid of StageStage if they specify **any** value for `OpenApi` (i.e only if they don't specify a value do we keep it).

File: samtranslator/swagger/swagger.py (hunk 5 above)
Comment (keetonian) [reply samtranslator/swagger/swagger.py]: Oh yeah, this will do string matching. `"3.0"` is inside of `"3.0.0"`. 👍

File: samtranslator/model/apigateway.py (hunk 4 above)
Comment (keetonian) [reply samtranslator/model/apigateway.py]: I think (but am not sure) that we still want this StageName property set, but not the other one that is on the `Deployment` resource. Either way, this will still work (we're still setting the StageName property on the Stage resource and conditionally removing it from the Deployment resource), so this is more of a technicality than a requirement.

File: samtranslator/model/api/api_generator.py (hunk 1 above)
Comment (keetonian) [reply samtranslator/model/api/api_generator.py]: That would work

File: samtranslator/model/apigateway.py (hunk 4 above)
Comment (brettstack) [reply samtranslator/model/apigateway.py]: I'd love to make this optional. Is there any reason not to?

File: samtranslator/model/apigateway.py (hunk 3 above)
Comment (praneetap) [reply samtranslator/model/apigateway.py]: will do thanks!

File: samtranslator/model/apigateway.py (hunk 4 above)
Comment (brettstack) [reply samtranslator/model/apigateway.py]: If we decide to make this optional, we need to decide on a default value ("prod"?) and update docs.

Comment (praneetap) [reply samtranslator/model/apigateway.py]: I think that should be a part of another cr. I don't want to stray too far from Sanath's changes for this bug previously.

Comment (praneetap) [reply samtranslator/model/apigateway.py]: I removed this change thanks Keeton!

File: samtranslator/model/api/api_generator.py (hunk 2 above)
Comment (praneetap) [reply samtranslator/model/api/api_generator.py]: Had an offline conversation about this, we will only fix stage stage if this flag is specified. If they don't have it specified and use oas 3 swagger in their api definitionbody it will still create an additional stage. This is because we want to remain backwards compatible and dont want to fix the stage stage bug without the customer opting in (with use of the OpenApiVersion flag)

File: samtranslator/swagger/swagger.py [hunk 6]
```diff
@@ -554,10 +555,15 @@ def is_valid(data):
         :param dict data: Data to be validated
         :return: True, if data is a Swagger
         """
-        return bool(data) and \
-            isinstance(data, dict) and \
-            bool(data.get("swagger")) and \
-            isinstance(data.get('paths'), dict)
+        openapi_version_supported_regex = r"\A3(\.\d)(\.\d)?$"
```
Comment (keetonian) [inline samtranslator/swagger/swagger.py]: Could you consolidate these regexes? It would be easier to maintain them in case we need to update the version in the future. I've seen two different ones, but one of them is used many more times than the other.

File: samtranslator/model/api/api_generator.py [hunk 7]
```diff
@@ -93,6 +97,11 @@ def _construct_rest_api(self):
             raise InvalidResourceException(self.logical_id,
                                            "Specify either 'DefinitionUri' or 'DefinitionBody' property and not both")
 
+        if self.open_api_version:
+            if re.match(OpenApiVersionsSupportedRegex, self.open_api_version) is None:
+                raise InvalidResourceException(
+                    self.logical_id, "The OpenApiVersion value must be of the format 2.x.x or 3.x.x")
```
Comment (brettstack) [inline samtranslator/model/api/api_generator.py]: "2.x.x" isn't valid in that regex, right? This is confusing because "2.x.x" is a commonly accepted value for semantic versions.

Comment (praneetap) [reply samtranslator/model/api/api_generator.py]: 2.x.x is a valid value for the flag. This is the regex - r"\A[2-3](\.\d)(\.\d)?$"

File: samtranslator/swagger/swagger.py (hunk 6 above)
Comment (praneetap) [reply samtranslator/swagger/swagger.py]: Where would the right place to consolidate be?

File: docs/cloudformation_compatibility.rst [hunk 8]
```diff
@@ -171,6 +171,7 @@ BinaryMediaTypes                    All
 MinimumCompressionSize              All
 Cors                                All
 TracingEnabled                      All
+OpenApiVersion                      All
```
[the remark this answers is not in the document]
Comment (praneetap) [reply docs/cloudformation_compatibility.rst]: Not sure if that is necessary? Can Brett comment here?

[the remark this answers is not in the document]
Comment (brettstack) [reply docs/cloudformation_compatibility.rst]: Preference is to support; otherwise we need to change this to "None"

File: samtranslator/model/api/api_generator.py (hunk 7 above)
Comment (brettstack) [reply samtranslator/model/api/api_generator.py]: 'x' isn't a digit/number though. Maybe there's some Python thing I'm missing here.

File: docs/cloudformation_compatibility.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (praneetap) [reply docs/cloudformation_compatibility.rst]: I am not sure yet whether this is necessary right now, so I am going to change this to None and we can tackle it in a separate task?

File: samtranslator/model/api/api_generator.py (hunk 7 above)
Comment (praneetap) [reply samtranslator/model/api/api_generator.py]: Got it, resolved. Thanks!

File: samtranslator/model/api/api_generator.py [hunk 9]
```diff
@@ -98,9 +97,9 @@ def _construct_rest_api(self):
                                            "Specify either 'DefinitionUri' or 'DefinitionBody' property and not both")
 
         if self.open_api_version:
-            if re.match(OpenApiVersionsSupportedRegex, self.open_api_version) is None:
+            if re.match(SwaggerEditor.get_openapi_versions_supported_regex(), self.open_api_version) is None:
                 raise InvalidResourceException(
-                    self.logical_id, "The OpenApiVersion value must be of the format 2.x.x or 3.x.x")
+                    self.logical_id, "The OpenApiVersion value must be 3.0.0")
```
Comment (brettstack) [inline samtranslator/model/api/api_generator.py]: "must be of the format 3.0.0" is fine; it's just that "2.x.x" was ambiguous.

Comment (praneetap) [reply samtranslator/model/api/api_generator.py]: done. thanks!

File: samtranslator/model/api/api_generator.py [hunk 10]
```diff
@@ -93,6 +96,11 @@ def _construct_rest_api(self):
             raise InvalidResourceException(self.logical_id,
                                            "Specify either 'DefinitionUri' or 'DefinitionBody' property and not both")
 
+        if self.open_api_version:
+            if re.match(SwaggerEditor.get_openapi_versions_supported_regex(), self.open_api_version) is None:
+                raise InvalidResourceException(
+                    self.logical_id, "The OpenApiVersion value must be of the format 3.0.0")
```
Comment (keetonian) [inline samtranslator/model/api/api_generator.py]: nit: `3.0.0` doesn't seem like a format, but I think this is still understandable.

**Comment by praneetap** [comment]
> Looks good. Have you deployed any of these OpenApi 3.0 templates to verify that they work in CFN?
@keetonian 
I could deploy this successfully
```
Globals:
  Api:
    OpenApiVersion: '3.0.1'

Resources:
  ImplicitApiFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: s3://praneep-test-2/thumbnails.zip
      Handler: index.gethtml
      Runtime: nodejs8.10
      Events:
        GetHtml:
          Type: Api
          Properties:
            Path: /
            Method: get
```
